File: lang/java/cxlib/src/test/java/cx/SchemaValidateTest.java
package cx;

import org.junit.jupiter.api.*;
import static org.junit.jupiter.api.Assertions.*;

import java.util.List;

/**
 * Smoke test for the Java schema-validator wrapper. Mirrors the
 * Rust binding's smoke set; the full 51-fixture conformance sweep
 * lives on the V/Python/Go side (same C ABI).
 */
public class SchemaValidateTest {

    private static final String BOOK_SCHEMA = """
            [?cx schema-of book]

            [book
              [body :elem]
              [attr id :string :req]
              [elem title :card='1..1']
              [elem author :card='1..*']
            ]

            [title [body :string]]
            [author [body :string]]
            """;

    @Test
    void validBookProducesNoDiagnostics() {
        String doc = """
                [book id='b1'
                  [title 'The Stand']
                  [author 'King']
                ]
                """;
        ValidationReport r = CxLib.validate(doc, BOOK_SCHEMA);
        assertTrue(r.isValid(), "expected zero errors, got " + r.errorCodes());
        assertEquals(0, r.errorCount());
    }

    @Test
    void missingRequiredAttrFiresS002() {
        String doc = """
                [book
                  [title 'X']
                  [author 'Y']
                ]
                """;
        ValidationReport r = CxLib.validate(doc, BOOK_SCHEMA);
        assertEquals(List.of("S002"), r.errorCodes());
        assertEquals(Severity.ERROR, r.diagnostics.get(0).severity());
    }

    @Test
    void rootMismatchFiresS017() {
        ValidationReport r = CxLib.validate("[other id='x']", BOOK_SCHEMA);
        assertEquals(List.of("S017"), r.errorCodes());
    }

    @Test
    void schemaDirectiveWithoutCallerSchemaFiresS010() {
        // spec/schema.md §13 — directive present but validator has no
        // way to resolve the path (no caller-supplied schema source).
        String doc = """
                [?cx schema=path/to/book.cxs]
                [book id='b1'
                  [title 'X']
                ]
                """;
        ValidationReport r = CxLib.validate(doc, "");
        assertEquals(List.of("S010"), r.errorCodes());
    }

    @Test
    void applyDefaultsWritesModifiedDoc() {
        String schema = """
                [?cx schema-of server]

                [server
                  [body :elem]
                  [attr host :string :def='localhost']
                ]
                """;
        ValidationReport r = CxLib.validateWithDefaults("[server]", schema);
        assertTrue(r.isValid(), "unexpected errors: " + r.errorCodes());
        assertTrue(r.modifiedDoc.contains("host="),
                "expected modifiedDoc to carry the default; got " + r.modifiedDoc);
    }
}

File: lang/rust/cxlib/tests/schema_validate_test.rs
//! Smoke test for the Rust schema-validator wrapper. Mirrors a few
//! fixtures from `conformance/schema_validate.txt`; the full conformance
//! sweep lives on the V/Python/Go side and exercises the same C ABI.

use cxlib::validate::{validate, validate_apply_defaults, Severity};

const BOOK_SCHEMA: &str = r#"
[?cx schema-of book]

[book
  [body :elem]
  [attr id :string :req]
  [elem title :card='1..1']
  [elem author :card='1..*']
]

[title [body :string]]
[author [body :string]]
"#;

#[test]
fn valid_book_no_diagnostics() {
    let doc = r#"
[book id='b1'
  [title 'The Stand']
  [author 'King']
]
"#;
    let report = validate(doc, BOOK_SCHEMA).expect("validate ok");
    assert!(report.is_valid(), "expected zero errors, got {:?}", report.error_codes());
    assert_eq!(report.error_count(), 0);
}

#[test]
fn missing_required_attr_fires_s002() {
    let doc = r#"
[book
  [title 'X']
  [author 'Y']
]
"#;
    let report = validate(doc, BOOK_SCHEMA).expect("validate ok");
    assert_eq!(report.error_codes(), vec!["S002".to_string()]);
    assert_eq!(report.diagnostics[0].severity, Severity::Error);
}

#[test]
fn root_mismatch_fires_s017() {
    let doc = "[other id='x']";
    let report = validate(doc, BOOK_SCHEMA).expect("validate ok");
    assert_eq!(report.error_codes(), vec!["S017".to_string()]);
}

#[test]
fn schema_directive_no_caller_schema_fires_s010() {
    // spec/schema.md §13 — directive present but validator has no
    // way to resolve the path (no caller-supplied schema source).
    let doc = r#"
[?cx schema=path/to/book.cxs]
[book id='b1'
  [title 'X']
]
"#;
    let report = validate(doc, "").expect("validate ok");
    assert_eq!(report.error_codes(), vec!["S010".to_string()]);
}

#[test]
fn apply_defaults_writes_modified_doc() {
    // Schema declares `:def='localhost'` for `host`; the doc omits it.
    // The defaults pass writes the populated form into modified_doc.
    // Sticking to a string-typed attribute here avoids the unrelated
    // i32-default-from-string coercion path (which is the
    // S011 territory tested by sv-039).
    let schema = r#"
[?cx schema-of server]

[server
  [attr host::string [default 'localhost']]
]
"#;
    let doc = "[server]";
    let report = validate_apply_defaults(doc, schema).expect("validate_apply_defaults ok");
    assert!(report.is_valid(), "unexpected errors: {:?}", report.error_codes());
    assert!(
        report.modified_doc.contains("host="),
        "expected modified_doc to carry the default; got {:?}", report.modified_doc,
    );
}
